src/System.Net.HttpListener/src/System/Net/Unix/ChunkedInputStream.cs

```diff
@@ -0,0 +1,188 @@
+        private ChunkStream _decoder;
+        private HttpListenerContext _context;
+        private bool _no_more_data;
+
+        private class ReadBufferState
+        {
+            public byte[] Buffer;
+            public int Offset;
+            public int Count;
+            public int InitialCount;
+            public HttpStreamAsyncResult Ares;
+            public ReadBufferState(byte[] buffer, int offset, int count,
+                        HttpStreamAsyncResult ares)
+            {
+                Buffer = buffer;
+                Offset = offset;
+                Count = count;
+                InitialCount = count;
+                Ares = ares;
+            }
+        }
+
+        public ChunkedInputStream(HttpListenerContext context, Stream stream,
+                        byte[] buffer, int offset, int length)
+                    : base(stream, buffer, offset, length)
+        {
+            _context = context;
+            WebHeaderCollection coll = (WebHeaderCollection)context.Request.Headers;
+            _decoder = new ChunkStream(coll);
+        }
+
+        public ChunkStream Decoder
+        {
+            get { return _decoder; }
+            set { _decoder = value; }
+        }
+
+        public override int Read([In, Out] byte[] buffer, int offset, int count)
+        {
+            IAsyncResult ares = BeginRead(buffer, offset, count, null, null);
+            return EndRead(ares);
+        }
+
+        public override IAsyncResult BeginRead(byte[] buffer, int offset, int count, AsyncCallback cback, object state)
+        {
+            if (_disposed)
+                throw new ObjectDisposedException(GetType().ToString());
+
+            if (buffer == null)
+                throw new ArgumentNullException(nameof(buffer));
+
+            int len = buffer.Length;
+            if (offset < 0 || offset > len)
+                throw new ArgumentOutOfRangeException(nameof(offset), SR.offset_out_of_range);
+
+            if (count < 0 || offset > len - count)
+                throw new ArgumentOutOfRangeException(nameof(count), SR.offset_out_of_range);
+
+            HttpStreamAsyncResult ares = new HttpStreamAsyncResult();
+            ares._callback = cback;
+            ares._state = state;
+            if (_no_more_data)
+            {
+                ares.Complete();
+                return ares;
+            }
+            int nread = _decoder.Read(buffer, offset, count);
+            offset += nread;
+            count -= nread;
+            if (count == 0)
+            {
+                // got all we wanted, no need to bother the decoder yet
+                ares._count = nread;
+                ares.Complete();
+                return ares;
+            }
+            if (!_decoder.WantMore)
+            {
+                _no_more_data = nread == 0;
+                ares._count = nread;
+                ares.Complete();
+                return ares;
+            }
+            ares._buffer = new byte[8192];
```

Comment:
I expect there's a lot of opportunity here to improve performance, e.g. not allocating a new 8K buffer for every read.